tqdm/_tqdm.py

```diff
@@ -734,6 +752,10 @@ def __iter__(self):
                                 else smoothing * delta_t / delta_it + \
                                 (1 - smoothing) * avg_time
 
+                        # Acquire locks if parallel bars
```

Comment:
@lrq3000 Ah yes, I follow this myself. Project-based learning is what I tell others.

But I'm asking specifically: how did you learn about Python internals, specifically? trial and error? profiling?